### PR TITLE
[ci] make ci idempotent, try two

### PR DESCRIPTION
**Pull request by danking**
Database idempotency is easy, use `IF NOT EXISTS`. User cleanup is hard because
the job isn't deterministic. Each attempt will produce a different password. You
cannot retrieve a password of an extant user.

The bulk of the changes are in service of creating a job whose attempts will
race to create the password. Only one attempt will win that race and pass its
password on to the `create_database_job`. Those jobs will race to create
databases and users and secrets, but all those operations are now
idempotent (`CREATE USER` uses `IF NOT EXISTS`; the secret is always the same).

### CODE REVIEW

**Comment by danking** [comment]
cc: @cseed on this PR I would appreciate your cursory review

**Comment by cseed** [comment]
This seems like the right approach.  The other option is to expose the password in the config which we are trying to avoid.

So ... even this is not idempotent.  The current version of batch doesn't guarantee that children (or multiple retries of the same child) get the same or consistent set of files from their parents.  This is because, while an attempt may have succeeded, there may be another attempt that is pending that succeeds and overwrites the outputs of the original successful attempt.

I fixed this in my google batch backend by storing attempt outputs in per-attempt directory: /path/to/scratch/files/job_id/attempt_id/file.  Then each child got the successful attempt (there could be multiple, but the driver selected one and only one) for each parent and that was used to localize the inputs.

So this good enough and we should plan to add attempt consistency to Batch in the future.

File: ci/ci/build.py [hunk 1]
```diff
@@ -876,38 +884,65 @@ def build(self, batch, code, scope):  # pylint: disable=unused-argument
...
+            secrets=[{
+                'namespace': self.database_server_config_namespace,
+                'name': 'database-server-config',
+                'mount_path': '/sql-config'
+            }],
+            service_account={
+                'namespace': BATCH_PODS_NAMESPACE,
+                'name': 'ci-agent'
+            },
+            output_files=[(x[1], x[0]) for x in password_files_input])
```
Comment (jigold) [inline ci/ci/build.py:929]: Should the parents of this job just be the deps_parents? Then the create_database job can just depend on this job.

Comment (danking) [reply ci/ci/build.py:929]: sure

**Comment by danking** [comment]
@jigold that's a great suggestion. To test that this code is acutally idempotent, I wanted to do this:

Duplicate the test database createDatabase task in `ci/test/resources/build.yaml` (but with a new step name). The jobs have the same parents so they race to run first. They have distinct passwords, so the resulting secret will non-deterministically have the wrong password (e.g. A creates the user, A writes its password first, B ignores already created user, then B writes its password second), but if everything but the secret works, then I'm confident repeated attempts sharing the same secret should work!

Unfortunately, I can't test CI itself in this way because test CIs (whether in dev or in a PR) are not permitted to create databases.

I took your suggestion and copied the SQL query out and ran it twice instead. That worked fine. Everything was created once.

**Comment by danking** [comment]
I also ran test_ci on this branch and it worked, but that doesn't tell us anything about the SQL for creating databases. The CI build page looks fine though, you see two jobs for each createDatabase step. The first one is suffixed `_create_passwords`

**Comment by danking** [comment]
bump. looks better with whitespace changes off: https://github.com/hail-is/hail/pull/8833/files?diff=unified&w=1

**Comment by danking** [comment]
be-rump

**Comment by danking** [comment]
bump @jigold

**Comment by jigold** [comment]
I think this is fine. I added the WIP tag. Take it off at your leisure in case things get messed up again.

**Comment by danking** [comment]
here we go!

**Comment by danking** [comment]
erm something is wrong with CI right now, let's not add to the crazy

**Comment by danking** [comment]
crazy time, go go go

**Pull request merged**